### PR TITLE
Handle highlight network failures

### PR DESCRIPTION
**Pull request by janKozie1**
for: https://github.com/openstax/unified/issues/905

### CODE REVIEW

**Comment by janKozie1** [comment]
It needs to wait for some changes from issue 1246 (tests in page.spec.tsx), but besides that, it should be ready for a review

File: src/app/notifications/reducer.spec.ts
Comment (PiotrKozlowski) [inline src/app/notifications/reducer.spec.ts:104]: Maybe you could try to mock the return value for `action.addToast(...)` to return another timestamp, this way we will avoid blocking a thread

File: src/app/notifications/types.ts
Comment (PiotrKozlowski) [inline src/app/notifications/types.ts:24]: maybe `messageKey` to indicate more clearly that this values will be used for localization?

File: src/app/utils.ts
Comment (PiotrKozlowski) [inline src/app/utils.ts:39]: just curious - is this change changing any behavior - it shouldn't right?
However, maybe it was written this way to make sure that `matches(action)` or `boundHook(action)` does not cause any side effects?

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
Looks good, I've left some minor comments

File: src/app/content/highlights/hooks/updateHighlight.ts
Comment (janKozie1) [reply src/app/content/highlights/hooks/updateHighlight.ts:21]: I'm going to wait until the final decision, since they are considering switching to one note for all errors.

File: src/app/utils.ts
Comment (janKozie1) [reply src/app/utils.ts:39]: The reason why I moved it is so that the action hooks get the access to store that was not yet updated (as long as they are sync which is why i removed async from `updateHighlight` for example).

from redux docs: `The last middleware in the chain will receive the real store's dispatch method as the next parameter`
`The store's reducing function will be called ... synchronously. Its return value will be considered the next state. It will be returned from getState() from now on`

I will double check if any of the existing hooks don't depend on the opposite thing happening

File: src/app/notifications/reducer.spec.ts
Comment (janKozie1) [reply src/app/notifications/reducer.spec.ts:104]: I'm just gonna reduce the timeout, the less mocks the better.

**Comment by janKozie1** [comment]
@PiotrKozlowski 
I'm still working on animating the toasts so anything related to that is not ready. If you want to wait with your review until I finish, I'm okay with that.

File: src/app/notifications/reducer.spec.ts
Comment (PiotrKozlowski) [reply src/app/notifications/reducer.spec.ts]: Maybe this can be used by using `jest.useFakeTimers();`? 
The thing is that we should try to make our unit tests as sync as possible, and `setTimeout` set even to `5`ms will block a thread for **at least** 5 ms. That sounds like a small amount of time but if we keep doing so we can end up a bigger and bigger amount

**Comment by PiotrKozlowski** [comment]
@janKozie1 ok, I'll wait

File: src/app/notifications/components/ToastNotifications/Toast.spec.tsx
Comment (PiotrKozlowski) [inline src/app/notifications/components/ToastNotifications/Toast.spec.tsx:154]: I think it would be okay to add to this test also the test case when `prevState.shouldAutoDismiss` is set to `true`

File: src/app/notifications/components/ToastNotifications/index.tsx
Comment (PiotrKozlowski) [inline src/app/notifications/components/ToastNotifications/index.tsx:22]: Should this be wrapped in `useMemo` so it's not recalculated on `mobileToolbaOpen` state change?

File: src/app/notifications/components/ToastNotifications/index.tsx
Comment (PiotrKozlowski) [inline src/app/notifications/components/ToastNotifications/index.tsx:47]: I feel like we should try to switch to the "modern" way when we can and use hooks for this

File: src/app/content/components/Page/PageComponent.tsx
Comment (PiotrKozlowski) [inline src/app/content/components/Page/PageComponent.tsx:7]: woops

File: src/app/content/components/Page/PageComponent.tsx
Comment (PiotrKozlowski) [inline src/app/content/components/Page/PageComponent.tsx]: woops

**Review by PiotrKozlowski — CHANGES_REQUESTED** [review]
at `0:55` on [this video](https://drive.google.com/file/d/1XQ4NmwggXVcOCdeimJLBzOnEngwh3sGr/view?usp=sharing) I've managed to break these toasts

Other comments are pretty minor things

File: src/app/content/components/Page/PageComponent.tsx
Comment (janKozie1) [reply src/app/content/components/Page/PageComponent.tsx]: :see_no_evil:

File: src/app/notifications/components/ToastNotifications/index.tsx
Comment (janKozie1) [reply src/app/notifications/components/ToastNotifications/index.tsx:22]: I'm not convinced that it would be faster since you won't get many errors to sort and memo comes with its own cost too

**Review by PiotrKozlowski — APPROVED** [review]

File: src/app/utils.ts
Comment (janKozie1) [reply src/app/utils.ts:40]: Turns out the pageView GA event depends on it, will need to rewrite the logic a bit.

Comment (PiotrKozlowski) [reply src/app/utils.ts]: Nice catch, please let me know when you finish fixing this so I can review

**Comment by janKozie1** [comment]
@PiotrKozlowski it should be ready for your review, thanks!

File: src/app/content/highlights/reducer.ts
Comment (PiotrKozlowski) [inline src/app/content/highlights/reducer.ts:131]: is `revertableHighlight` used for something?

**Review by PiotrKozlowski — COMMENTED** [review]
I've left minor comments, but there is also this one:
I think that maybe we should stop using `hooks` for creating / updating / deleting highlight and instead go with a [redux-thunk](https://github.com/reduxjs/redux-thunk) approach
In this case, you wouldn't have to pass `preUpdateData` and call any revert actions - instead, you would first call the api and then depends on the response, dispatch an action for creating a highlight in case of success or adding a toast in case of failure

File: src/app/content/highlights/components/EditCard.tsx
Comment (janKozie1) [reply src/app/content/highlights/components/EditCard.tsx:146]: I left it since old code was written that way. I think it's a good idea to change it.

**Comment by janKozie1** [comment]
While it would be nice to not need to send data require to revert changes, it would also mean that users need to wait for the request to finish to see the results of their actions so It would make the highlight-related UI at least a bit less responsive. 

It would also add a separate place where actions / network requests are handled which I don't think is a bad thing but would add more to this PR (including the types that would need to account for that change)

It could also prove to be problematic when users try to make highlights while the request is not finished yet (I'm mainly thinking about the discard note since `hasUnsavedChanges` depends on the action hitting the reducer).

File: src/app/content/highlights/components/EditCard.tsx
Comment (janKozie1) [reply src/app/content/highlights/components/EditCard.tsx]: I've extracted it and wrote tests for it but it wouldn't make the tests more readable in `EditCard.spec.tsx`

**Review by PiotrKozlowski — APPROVED** [review]

**Comment by PiotrKozlowski** [comment]
From the other hand, with redux-thunk we could do a function that dispatches create/update/delete action imidietally, then call the api and then dispatch revert action on failure.
Since this function would be called synchronously, we would have an access to the current state so we wouldn't have to pass anything new and we wouldn't have to chceck if this is reverting action.

But I agree that this is probably not the best moment to introduce a new approach

**Pull request merged**
